Layer 0 Spec: Module & Module Source

caridy opened the pull request:
This PR define the specs for [layer 0](https://github.com/tc39/proposal-compartments/blob/master/0-module-and-module-source.md).

The readable version of this PR can be found [here](https://raw.githack.com/tc39/proposal-compartments/103b378cd320fc38bf3b381c2a729dce49effec2/0-module-and-module-source.html).

### Details

1. Introduces `Module`
2. Introduces `ModuleSource`
3. Modifies Runtime Semantics for `import()` to accept a `Module` instance

Jack-Works commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,614 @@
+<emu-clause id="sec-modulesource-records">
+    <h1>Module Source Records</h1>
+
+    <p>A <dfn id="modulesource-record" variants="Module Source Records">Module Source Record</dfn> is used to represent information about a module source that was defined from ECMAScript source text (<emu-xref href="#sec-ecmascript-language-source-code"></emu-xref>) that was parsed using the goal symbol |Module|. Its fields contain digested information about the names that are imported and export ...
```
```html
(<emu-xref href="#sec-ecmascript-language-source-code"></emu-xref>)
```

this part is rendered as `()` in html.

caridy replied:
thanks for the help, biblio was missing.

Jack-Works commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,614 @@
+    <emu-clause id="sec-modulesource-abstracts">
+        <h1>ModuleSource Abstract Operations</h1>
+
+        <emu-clause id="sec-parsemodulesource" type="abstract operation">
```
This algr looks very similar to [ParseModule](https://tc39.es/ecma262/#sec-parsemodule). Can you use `<ins>` and `<del>` to indicate the changes?

caridy replied:
I have added a note below this operation related to this. I don't it will be a modifcation, but a refactor that splits ParseModule into two operations.

Jack-Works commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,614 @@
+                1. If Type(_importMeta_) is not Object, throw a *TypeError* exception.
+                1. If IsCallable(_importHook_) is *false*, throw a *TypeError* exception.
+                1. Perform ? RequireInternalSlot(_moduleSource_, [[ModuleSource]]).
+                1. Let _P_ be MakeBasicObject(&laquo; [[ImportHook]], [[ImportMeta]] &raquo;).
```
This variable is unused.

Jack-Works commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,614 @@
+                  1. return ? Call(_importHook_, *undefined*, « _specifier_, _importMeta_ »).
+                1. Set _O_.[[Module]] to _moduleRecord_.
+                1. Set _O_.[[ModuleSourceInstance]] to _moduleSource_.
+                1. Set _O_.[[ImportHook]] be CreateBuiltinFunction(_importHookClosure_, 0, *""*, &laquo; &raquo;).
```
```suggestion
                1. Set _O_.[[ImportHook]] be CreateBuiltinFunction(_importHookClosure_, 1, *""*, &laquo; &raquo;).
```

Jack-Works commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,614 @@
+                1. Assert: _module_ is a Module Record.
+                1. Assert: _module_.[[ImportMeta]] is ~empty~.
+                1. Let _importMeta_ to OrdinaryObjectCreate(*null*).
+                1. Let _keys_ be ? <emu-meta effects="user-code">_from_.[[OwnPropertyKeys]]()</emu-meta>.
```
Can we do `? Call (%Object.assign%, *null*, < _importMeta_, _meta_ >)`?

caridy replied:
being more conservative here since we don't use `Object.assign` anywhere.

Jack-Works commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,614 @@
+                  1. If _desc_ is not *undefined* and _desc_.[[Enumerable]] is *true*, then
+                    1. Let _nextValue_ be ? Get(_from_, _nextKey_).
+                    1. Perform ! CreateDataPropertyOrThrow(_importMeta_, _nextKey_,_nextValue_).
+                1. Perform HostFinalizeImportMeta(_importMeta_, _module_).
```
Why it will call the host hook? This will prevent virtualization if the host did something observable.

caridy replied:
`HostFinalizeImportMeta`: https://tc39.es/ecma262/#sec-hostfinalizeimportmeta

> Most [hosts](https://tc39.es/ecma262/#host) will be able to simply define [HostGetImportMetaProperties](https://tc39.es/ecma262/#sec-hostgetimportmetaproperties), and leave HostFinalizeImportMeta with its default behaviour. However, HostFinalizeImportMeta provides an "escape hatch" for [hosts](https://tc39.es/ecma262/#host) which need to directly manipulate the object before it is exposed to ECMAScript code.

Basically, the creation of the meta object involves two steps, one of them needs virtualization (`HostGetImportMetaProperties`), the other one `HostFinalizeImportMeta` doesn't.

Jack-Works replied:
No, since it does not call the `HostGetImportMetaProperties`, it should not call  `HostFinalizeImportMeta` either.

Jack-Works commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,614 @@
+              1. IfAbruptRejectPromise(_completion_, _promiseCapability_).
+              1. Let _importHookPromise_ be ? PromiseResolve(%Promise%, _completion_.[[Value]]).
+              1. Let _fulfilledClosure_ be a new Abstract Closure with parameters (_result_) that captures _moduleRecord_ and _promiseCapability_ and performs the following steps when called
+                1. Assert: Evaluate has already been invoked on _moduleRecord_ and successfully completed.
```
this does not look correct to me, it will prevent load nested dependencies.

nicolo-ribaudo replied:
+1, Evaluate cannot have been called if we are still loading the module dependencies.

nicolo-ribaudo commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,613 @@
+              1. Let _rejectedClosure_ be a new Abstract Closure with parameters (_error_) that captures _promiseCapability_ and performs the following steps when called:
+                1. Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; _error_ &raquo;).
+                1. Return ~unused~.
+              1. Let _onRejected_ be CreateBuiltinFunction(_rejectedClosure_, 0, *""*, &laquo; &raquo;).
```
```suggestion
              1. Let _onRejected_ be CreateBuiltinFunction(_rejectedClosure_, 1, *""*, &laquo; &raquo;).
```

nicolo-ribaudo commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,613 @@
+            <emu-alg>
+              1. Let _fulfilledClosure_ be a new Abstract Closure with parameters (_result_) that captures _moduleRecord_ and _promiseCapability_ and performs the following steps when called:
+                1. Assert: _result_ is *undefined*.
+                1. Assert: Evaluate has already been invoked on _moduleRecord_ and successfully completed.
```
Doesn't `importHook` return unlinked modules?

caridy commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,661 @@
+              1. Let _importHookPromise_ be ? PromiseResolve(%Promise%, _completion_.[[Value]]).
+              1. Let _fulfilledClosure_ be a new Abstract Closure with parameters (_result_) that captures _moduleRecord_ and _promiseCapability_ and performs the following steps when called
+                1. If _result_ is an Object that has a [[ImportHook]] internal slot, then
+                  1. Let _moduleRecord_ be _result_.[[Module]].
```
@nicolo-ribaudo @Jack-Works, I have two issues here:

1. if the result of calling `[[ImportHook]]` function must always be a Module Instance, then how can we resolve to `fs` or some other module instance that delegates the resolution to the browser? That seems like a problem. I don't know if the issue is here, or somewhere else, but definitely needs to be addressed.
2. where do we go into the recursive resolution? in `HostImportModuleRecordDynamically` we only talk about the "unique [ModuleSpecifier](https://tc39.es/ecma262/#prod-ModuleSpecifier) associated to the moduleRecord", never about the transitive dependencies. any suggestion?

guybedford replied:
One option here might be to support having import hook return a `ModuleSource` which is then instanced against its primary default instance for the host context? Would need to think through the edge cases though....

Recursive resolution happens through `HostResolveImportedModule` calls in `InnerModuleLinking` and `InnerModuleEvaluation`. I really liked @nicolo-ribaudo's approach in https://nicolo-ribaudo.github.io/modules-import-hooks-refactor/ of having this being a sync lookup against the resolution list on the module record itself though to more clearly separate the sync and async iterations.

nicolo-ribaudo replied:
1. My hope is that import reflection could also give `Module`s other than `ModuleSource`s <small>(the syntax is just an idea)</small>:
   ```js
   import module jsM from "x.js";
   import source jsS from "x.js";
   import module wasmM from "y.wasm";
   import source wasmS from "y.wasm";

   jsM instanceof Module;
   jsS instanceof ModuleSource;
   wasmM instanceof Module;
   wasmS instanceof WebAssembly.Module; // WebAssembly.Module respects the "module source interface", whatever that means

   jsM.source === jsS;
   wasmM.source === wasmS;
    ```

   Then, an import hook might be implemented like this:
   ```js
   function importHook(specifier, importMeta) {
     if (specifier === "fs") return import("fs", { reflection: "module" });
     
     const source = // get the module source
     const innerImportMeta = // get the import.meta
     return new Module(source, importHook, innerImportMeta);
   }
   ```
2. As @guybedford said, the current ECMA-262, we explicitly handle the recursive resolution: we hide the fact that "real resolution" can be asynchronous, and we pretend that the resolution happens in `HostResolveImportedModule`. We call it recursively for every transitive dependency in `InnerModuleLinking`, which is triggered by an (implicit) call to `.Link()` of the top-level imported module in `HostImportModuleDynamically`.
   We probably can keep hand waving in this proposal spec (we need a "complete spec text" only when asking for stage 3, and this PR is already very good for asking for stage 2), but a proper solution will look a lot like https://nicolo-ribaudo.github.io/modules-import-hooks-refactor/#sec-LoadRequestedModules (or, I push for merging https://nicolo-ribaudo.github.io/modules-import-hooks-refactor without waiting for a proposal, and then you just edit that in this proposal to call `[[ImportHook]]` instead of `HostLoadImportedModule`)

guybedford replied:
@nicolo-ribaudo on (1), this is an interesting proposal although does introduce more complexity. That said it very much resolves some current concerns through the separation. Would the module instance reflection be unexecuted if it hasn't already been executed? But would the dependencies be fully preloaded? If so, that may well make it suitable for deferred evaluation mechanisms....

Furthermore, passing an instance to another realm in this way with linkage might enable an easy solution to resolution sharing without having to re-instrument a loader on the other realm...

Would be interested to hear more details if it might be able to address the above.

guybedford commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,661 @@
+        1. <ins>Let _specifierOrModule_ be ? GetValue(_argRef_).</ins>
+        1. Let _promiseCapability_ be ! NewPromiseCapability(%Promise%).
+        1. <ins>If Type(_specifierOrModule_) is Object that has a [[ModuleSourceInstance]] internal slot, then</ins>
+          1. <ins>Let _moduleRecord_ be _specifierOrModule_.[[Module]].</ins>
```
Also, should we also be looking up if `_referencingScriptOrModule_` is itself a `SourceTextModule` (pretty sure we can assume it always be?) that we should look up its `[[Instance]]` slot in turn for the parent reference if there is one? Would that be useful information for an import hook even at the top level?

caridy replied:
> Should this be `[[Module]]` or `[[ModuleSourceInstance]]`?

It should be `[[Module]]` because the instance provided must already have a qualifying module record associate to it, and that's the currency of all these operations.

caridy replied:
> We may want to make that invariant explicit.

Any suggestion on how to make that more explicit?

guybedford commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,661 @@
+              1. Assert: ResolveModuleRecordDependency has not been invoked with _moduleRecord_ and _specifier_ pair.
+              1. Let _moduleInstance_ be _moduleRecord_.[[ModuleInstance]].
+              1. Let _importHook_ be _moduleInstance_.[[ImportHook]].
+              1. Let _completion_ be Completion(Call(_importHook_, *undefined*, « _specifier_ »).
```
If we want to ensure sync resolution possibilities, perhaps we do need to support `Promise<Instance> | Instance` as the API rather here?

caridy replied:
that is tricky as always... because the instance might have a `.then` on them, and then what?

Jack-Works replied:
Pass the result to `PromiseResolve(_here_)` and it gets unified? 🤔

guybedford replied:
Can we define the `Instance` JS object to explicitly never have a `then` property?

caridy replied:
> Can we define the Instance JS object to explicitly never have a then property?

@guybedford  No, I don't think we should do that. In fact `Module` is subclass-able.

> Pass the result to PromiseResolve(_here_) and it gets unified?

@Jack-Works that's precisely what I do in the following steps.

guybedford reviewed (commented):
Seems roughly on the right track to me.

nicolo-ribaudo commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,661 @@
+            <p>An implementation of HostImportModuleRecordDynamically must conform to the following requirements:</p>
+    
+            <ul>
+              <li>
+                It must return ~unused~. It must call ResolveModuleRecordDependency for every unique |ModuleSpecifier| associated to the _moduleRecord_. Success or failure must instead be signaled as discussed below.
+              </li>
```
```suggestion
              <li>
                It must call ResolveModuleRecordDependency for every (_moduleRecord_, _specifierString_) pair, where _specifierString_ are the entries of _moduleRecord.[[RequestedModules]] for each recursively resolved Cyclic Module Record _moduleRecord_. ResolveModuleRecordDependency must not be called twice with the same (_moduleRecord_, _specifierString_) pair: instead, the result of the first call should be cached and reused as the result of successive resolutions.
              </li>
              <li>
                It must return ~unused~. Success or failure must instead be signaled as discussed below.
              </li>
```

This requests the host to to the process transitively to load all the dependencies.

caridy commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,661 @@
+        </tr>
+        <tr>
+          <td>
+            ...
```
no deletions yet, that can come later. For now, there is a new entry in this record, and a new similar record (Module Source Record).

caridy commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,661 @@
+</emu-clause>
+
+<emu-clause id="sec-modulesource-records">
+    <h1>Module Source Records</h1>
```
We can do that later. The idea is to modify the minimum of the current spec to expedite layer 0. So, it is intentional. I don't think it belongs to the Cyclic hierarchy because this is not really a module record, this is just static parsing information. Maybe it should be just an internal slot of the Module Record.

guybedford replied:
Sure that's fine for now.

erights reviewed (changes requested):
At https://raw.githack.com/tc39/proposal-compartments/e897e4a5707154f4af67ac89ec7bf7cfef966636/0-module-and-module-source.html 4.1 step 7 I see

<img width="888" alt="image" src="https://user-images.githubusercontent.com/273868/181404203-2f360482-04c1-4b08-b0f6-96ff72f9f0c3.png">

If 7.b completes abruptly, shouldn't 7.c be skipped?

erights reviewed (commented):
Should https://raw.githack.com/tc39/proposal-compartments/e897e4a5707154f4af67ac89ec7bf7cfef966636/0-module-and-module-source.html specify that `ModuleSource.prototype.constructor` is the `ModuleSource` constructor? Or is that taken care of by a generic preamble?

erights commented:
Step 3
> 3. If [Type](https://tc39.es/ecma262/#sec-ecmascript-data-types-and-values)(importHook) is not Callable, throw a TypeError exception.

seems redundant with step 6
> 6. If [IsCallable](https://tc39.es/ecma262/#sec-iscallable)(importHook) is false, throw a TypeError exception.

Jack-Works commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,639 @@
+    
+            <ul>
+              <li>
+                It must call ResolveModuleRecordDependency for every (_moduleRecord_, _specifierString_) pair, where _specifierString_ are the entries of _moduleRecord.[[RequestedModules]] for each recursively resolved Cyclic Module Record _moduleRecord_. ResolveModuleRecordDependency must not be called twice with the same (_moduleRecord_, _specifierString_) pair: instead, the result of the first ...
```
```suggestion
                It must call ResolveModuleRecordDependency for every (_moduleRecord_, _specifierString_) pair, where _specifierString_ are the entries of _moduleRecord_.[[RequestedModules]] for each recursively resolved Cyclic Module Record _moduleRecord_. ResolveModuleRecordDependency must not be called twice with the same (_moduleRecord_, _specifierString_) pair: instead, the result of the first call should be cached and reused as the result of successive resolutions. The resolution for a given pair must not be result of the same Module Record.
```

Jack-Works replied:
> The resolution for a given pair must not be result of the same Module Record.

Why? A module can import itself.

caridy replied:
really? cc @erights

kriskowal replied:
Verified empirically! In Node.js (using 14 for the moment):

```js
import { name as nom } from './self.mjs';
export const name = 10;
```

caridy replied:
alright! in that case, I will update the text to remove that wording! that was surprising.

nicolo-ribaudo commented:
@erights 
> If 7.b completes abruptly, shouldn't 7.c be skipped?

`IfAbruptRejectPromise` has an early return if the completion is an abrupt completion, so 7.c is skipped.

caridy commented:
@erights 

> Should https://raw.githack.com/tc39/proposal-compartments/e897e4a5707154f4af67ac89ec7bf7cfef966636/0-module-and-module-source.html specify that `ModuleSource.prototype.constructor` is the `ModuleSource` constructor? Or is that taken care of by a generic preamble?

it was missing for both, I just added.

> Step 3
> 
> > 3. If [Type](https://tc39.es/ecma262/#sec-ecmascript-data-types-and-values)(importHook) is not Callable, throw a TypeError exception.
> 
> seems redundant with step 6
> 
> > 6. If [IsCallable](https://tc39.es/ecma262/#sec-iscallable)(importHook) is false, throw a TypeError exception.

Fixed.

naugtur commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,648 @@
+                    <tr>
+                        <td>[[ModuleSourceInstance]]</td>
+                        <td>ModuleSource Instance or *null*</td>
+                        <td>The Source Text Module Record that represents this Module Instance. *null* if this Module Instance has not a source associated to it.</td>
```
What's the purpose of having [[Module]] and [[ModuleSourceInstance]] with the same description but one of them can be null while the other can't? 
This looks like a mistake or a convention I misunderstood.

caridy replied:
Certainly a mistake in the description. I can fix the description of `[[ModuleSourceInstance]]` to be:

"The ModuleSource Instance associated to Module Instance. *null* if this Module Instance has not a source associated to it."

As for the why this one can be null, there is a note few lines below in the emu-note that describes the scenario where this can be null with an example.

naugtur commented:
I made an attempt to visualize this change for the benefit of my own mental model. Sharing in case it's useful:
https://gist.github.com/naugtur/daf2aaf415e03e6c3ae5030e06f7b20e

caridy commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,648 @@
+                    <tr>
+                        <td>[[ModuleSourceInstance]]</td>
+                        <td>ModuleSource Instance or *null*</td>
+                        <td>The Source Text Module Record that represents this Module Instance. *null* if this Module Instance has not a source associated to it.</td>
```
```suggestion
                        <td>The ModuleSource Instance associated to Module Instance. *null* if this Module Instance has not a source associated to it.</td>
```

caridy commented:
> I made an attempt to visualize this change for the benefit of my own mental model. Sharing in case it's useful: https://gist.github.com/naugtur/daf2aaf415e03e6c3ae5030e06f7b20e

Very nice, I added a note in the gist.

nicolo-ribaudo commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,650 @@
+        </ul>
+
+        <emu-clause id="sec-module">
+            <h1>Module ( _moduleSource_, _referral_, _importHook_ [, _importMeta_ ] )</h1>
```
We have always discussed about a _string_ referrer, but this spec text allows `referral` to be any ECMAScript value. I very much prefer this behavior, rather than restricting to strings (I was going to open a new issue proposing to relax the restriction, before noticing that it's already not restricted).

I realized that passing the referrer to `importHook` (and thus passing `referral` to `new Module`) is not necessary, but it's just an ergonomic improvement. If the spec didn't call `importHook` passing the referrer, I could just do this:
```js
function buildImportHook(referrer) {
  return async function importHook(specifier) {
    const url = new URL(specifier, referrer);
    const sourceText = await fetch(url).then(r => r.text());
    const source = new ModuleSource(sourceText);
    return new Module(source, buildImportHook(url));
  };
};

const entryModule = new Module(entrySource, buildImportHook(entryURL));
await import(entryModule);
```
(or I could use `.bind` rather than manual currying)

Passing the `referrer` through the `Module` constructor to `importHook` is just an ergonomic/performance improvement to avoid creating closures that capture context: it's similar to the `thisArg` parameter of all the array methods that take a function.

Since we are "just" improving ergonomics, we can look at other use cases for providing some context to the `importHook` function. One example ([taken from the HTML spec](https://html.spec.whatwg.org/#script-fetch-options)) is to provide some "fetch options" that describe how the dependencies should be fetched:
```js
async function importHook(specifier, { baseURL, referrerPolicy }) {
  const url = new URL(specifier, baseURL);
  const sourceText = await fetch(url, { referrerPolicy }).then(r => r.text());
  const source = new ModuleSource(sourceText);
  return new Module(source, {
    baseURL: url,
    referrerPolicy,
  });
}

const entryModule = importHook("./main.js", { baseURL: document.URL, referrerPolicy: "same-origin" });
await import(entryModule);

// That is roughly equivalent to
//   <script type="module" src="./main.js" referrerpolicy="same-origin" />
```

Allowing any object aligns this "context parameter" with the current [[HostDefined]] slot on Module Records: it's a general-purpouse place used to share loader-level information between modules.

So :+1: to not checking that `Type(_referral_)` is string, and maybe we could rename it to `_context_` (or `_thisArg_` and pass it as `this` to `importHook`, like array methods do).

caridy replied:
@nicolo-ribaudo  in principle I agree with this. Question: in the example above I suspect you are missing `importHook`  argument when calling `new Module()` right?

nicolo-ribaudo replied:
Yes, fixed!

caridy replied:
Resolution: We will look into this in another PR.

Jack-Works commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,650 @@
+        </ul>
+
+        <emu-clause id="sec-module">
+            <h1>Module ( _moduleSource_, _referral_, _importHook_ [, _importMeta_ ] )</h1>
```
`importHook` should be optional argument. Inherit the current evaluator's import hook.

```js
const x = new Module(source, 'spec') // uses host
const ev = new Evaluators({ importHook: myHook })
new ev.Module(source, 'spec') // uses myHook
```

caridy replied:
Resolution: we will look into the default hook in another PR/issue.

caridy commented:
@kriskowal feel free to merge this. It seems that we are all in agreement according to the last SES meeting.

Jack-Works commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,650 @@
+            <emu-alg>
+                1. If NewTarget is *undefined*, throw a *TypeError* exception.
+                1. Let _O_ be ? OrdinaryCreateFromConstructor(NewTarget, *"%ModuleSource.prototype%"*, « [[ModuleSource]] »).
+                1. Let _body_ be ParseText(_sourceText_, |Module|).
```
It should call `? HostEnsureCanCompileStrings(_currentRealm_)` before ParseText.

caridy replied:
This is the wrong place to do so... creating an instance of module source should be possible even when CSP prevents evaluation of code. Perhaps the host hook has the wrong name... it is not really about compilation, but evaluation. I'm fixing this in https://github.com/tc39/proposal-compartments/pull/76.

Jack-Works commented on 0-module-and-module-source.emu:
```diff
@@ -0,0 +1,650 @@
+            <emu-alg>
+                1. If NewTarget is *undefined*, throw a *TypeError* exception.
+                1. Let _O_ be ? OrdinaryCreateFromConstructor(NewTarget, *"%ModuleSource.prototype%"*, « [[ModuleSource]] »).
+                1. Let _body_ be ParseText(_sourceText_, |Module|).
```
This step should do a coherent to convert _sourceText_ to a String.

caridy replied:
fixed by https://github.com/tc39/proposal-compartments/pull/76

kriskowal commented:
I believe “referrer” is the more correct word-choice than “referral”, but that is not important yet.
